### PR TITLE
RFC: Add SetRuntimeEnvironment call for NI-DCPower

### PR DESCRIPTION
**Pull request by ni-jfitzger**
### What does this Pull Request accomplish?

Add an NiDCPowerLibrary::SetRuntimeEnvironment call upon library load, before init is called.

Implemented via the following changes:
* Update `source/codegen/metadata/nidcpower/functions_addon.py` to privately add the SetRuntimeEnvironment function
* Update `source/codegen/metadata/nidcpower/__init__.py` so that the new metadata will be merged. I merged metadata for attributes, functions and enums. I didn't mess with config_addon.py, though.
* Update library.cpp.mako to call SetRuntimeEnvironment if the function exists
* Generate a version.h with some of the values found in version.rc, defined as string constants (couldn't figure out any other way to get access to the data easily)
* Update CMakeLists.txt and generate

### Why should this Pull Request be merged?

As developers, we want to know how much nigrpc_device_server is used and which versions.

### What testing has been done?

* Built ni_grpc_device_server.exe locally and called it with the nimi-python system tests in CEIP test mode.
   * Verified that the Runtime Environment data was logged correctly.
* Manually launched the server and called it, opening and closing sessions, as well as the python process that I called it from, multiple times. Confirmed that data was only logged once. This makes sense because the service that loads the library is supposed to remain loaded for the lifetime of the server instance.

TODOs:
* Before submitting this, merge changes from main after #964 is submitted and add a unit test. I might just close this and create a separate PR.

### CODE REVIEW

**Comment by ni-jfitzger** [comment]
TODO: update tests? Maybe testing should be done with nifake?

File: source/codegen/templates/library.cpp.mako [hunk 1]
```diff
@@ -44,6 +45,10 @@ ${service_class_prefix}Library::${service_class_prefix}Library() : shared_librar
 %>\
   function_pointers_.${method_name} = reinterpret_cast<${method_name}Ptr>(shared_library_.get_function_pointer("${c_name}"));
 % endfor
+% if 'SetRuntimeEnvironment' in service_helpers.filter_api_functions(functions, only_mockable_functions=False):
+
+  this->SetRuntimeEnvironment(nidevice_grpc::NIDEVICE_GRPC_ORIGINALFILENAME.c_str(), nidevice_grpc::NIDEVICE_GRPC_FILEVERSION.c_str(), "", "");
```
Comment (ni-jfitzger) [inline source/codegen/templates/library.cpp.mako:50]: Maybe I need to do a try, catch in case this errors?

File: source/codegen/metadata/nidcpower/__init__.py [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
 metadata = {
     "functions" : functions,
     "attributes" : attributes,
     "enums" : enums,
     "config" : config
-}
\ No newline at end of file
+}
+
+metadata['functions'].update(functions_override_metadata)
```
Comment (ni-jfitzger) [inline source/codegen/metadata/nidcpower/__init__.py:16]: Perhaps this is too simplistic a method of merging?
I considered borrowing from nimi-python's codegen.

**Comment by ni-jfitzger** [comment]
I'm not sure why the validate_python check failed. I don't think it's related to my changes.

File: source/codegen/templates/library.cpp.mako (hunk 1 above)
Comment (SparkingSpork) [reply source/codegen/templates/library.cpp.mako:50]: Rather than adding specific logic for a specific function, it might be better to define this requirement in the metadata and use that to add any extra "initialization functions" so that this is extensible to other functions in the future. I'll defer to @reckenro or @astarche on that since they know more about the code generation, though.

File: CMakeLists.txt [hunk 3]
```diff
@@ -425,6 +425,17 @@ if(WIN32)
   )
 endif()
 
+# Generate version header file
```
Comment (SparkingSpork) [inline CMakeLists.txt:428]: Can you add a comment that this is done via CMake so that it uses version variables from the CMake environment which aren't directly available to our other code generation? That way it's clear why this particular file isn't generated via mako.

**Review by SparkingSpork — APPROVED** [review]

**Comment by bkeryan** [comment]
> I'm not sure why the validate_python check failed. I don't think it's related to my changes.

black 23.7 was released a week ago and it dropped Python 3.7 support.

validate_examples.py generates temporary Poetry projects on the fly, which is kind of unusual. It sounds like `poetry add` (at least in Poetry 1.2.2) decides to depend on black = "^23.7.0" before it realizes this won't work with Python 3.7. The easiest solution is probably to upgrade the Python version for the GitHub workflow to 3.8 or later.

Cc: @mshafer-NI

File: generated/version.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#ifndef NIDEVICE_GRPC_VERSION_H
+#define NIDEVICE_GRPC_VERSION_H
+
+#include <string>
+
+namespace nidevice_grpc {
+
+std::string const NIDEVICE_GRPC_MAJORNUMBER = "2";
```
Comment (bkeryan) [inline generated/version.h:8]: Defining static STL objects in a header file duplicates them in every translation unit that includes the header.

Comment (bkeryan) [reply generated/version.h:8]: Option 1: use C strings

version.h:
```
const char* const NIDEVICE_GRPC_MAJORNUMBER = "2";
```

Option 2: use extern STL strings

version.h:
```
extern const std::string NIDEVICE_GRPC_MAJORNUMBER;
```

version.cpp
```
const std::string NIDEVICE_GRPC_MAJORNUMBER = "2";
```

File: generated/nidcpower/nidcpower_library.h [hunk 5]
```diff
@@ -166,6 +166,7 @@ class NiDCPowerLibrary : public nidcpower_grpc::NiDCPowerLibraryInterface {
   ViStatus UnlockSession(ViSession vi, ViBoolean* callerHasLock);
   ViStatus WaitForEvent(ViSession vi, ViInt32 eventId, ViReal64 timeout);
   ViStatus WaitForEventWithChannels(ViSession vi, ViConstString channelName, ViInt32 eventId, ViReal64 timeout);
+  ViStatus SetRuntimeEnvironment(ViConstString environment, ViConstString environmentVersion, ViConstString reserved1, ViConstString reserved2);
```
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower_library.h:169]: You should consider adding `override` to all these, but in its own PR.

File: source/codegen/metadata/nidcpower/functions_addon.py [hunk 6]
```diff
@@ -2,4 +2,38 @@
 # Changes to the metadata should be made here, because functions.py is generated thus any changes get overwritten.
 
 functions_override_metadata = {
```
Comment (marcoskirsch) [inline source/codegen/metadata/nidcpower/functions_addon.py:4]: I think updating metadata should be done in its own PR. Can you do that?

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: source/server/version.h.in [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#ifndef NIDEVICE_GRPC_VERSION_H
+#define NIDEVICE_GRPC_VERSION_H
+
+#include <string>
+
+namespace nidevice_grpc {
+
+std::string const NIDEVICE_GRPC_MAJORNUMBER = "@PROJECT_VERSION_MAJOR@";
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/server/version.h.in:8]: Mostly, snake_case is used.
Although, all the global const strings I see are all `static const char*` with kCamelCase names. I've seen the convention before, but don't really know it's intended use. Does the k stand for kernel?

[the remark this answers is not in the document]
Comment (bkeryan) [reply source/server/version.h.in:8]: 'k' is for konstant.

File: source/server/version.h.in [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#ifndef NIDEVICE_GRPC_VERSION_H
+#define NIDEVICE_GRPC_VERSION_H
+
+#include <string>
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/server/version.h.in:4]: Hmm, yeah, I might even be able to get away with deleting the .cstr() calls with this.
I'm guessing I don't need extern to prevent code bloat?

[the remark this answers is not in the document]
Comment (bkeryan) [reply source/server/version.h.in:4]: `string_view` is not NUL-terminated and doesn't have a `.c_str()` method. To pass a `string_view` to an API that expects a C string, you should create a temporary `std::string` and call `.c_str()`, which is more complicated than using a constant C string in the first place.

File: generated/version.h (hunk 4 above)
Comment (bkeryan) [reply generated/version.h:8]: `constexpr` C string pointers are also fine:
```
constexpr const char* NIDEVICE_GRPC_MAJORNUMBER = "2";
```

File: generated/nidcpower/nidcpower_library.cpp [hunk 9]
```diff
@@ -1366,4 +1370,12 @@ ViStatus NiDCPowerLibrary::WaitForEventWithChannels(ViSession vi, ViConstString
   return function_pointers_.WaitForEventWithChannels(vi, channelName, eventId, timeout);
 }
 
+ViStatus NiDCPowerLibrary::SetRuntimeEnvironment(ViConstString environment, ViConstString environmentVersion, ViConstString reserved1, ViConstString reserved2)
+{
+  if (!function_pointers_.SetRuntimeEnvironment) {
+    throw nidevice_grpc::LibraryLoadException("Could not find niDCPower_SetRuntimeEnvironment.");
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply generated/nidcpower/nidcpower_library.cpp:1376]: The wrapper function checks if the function loaded. I can do an if statement to do the same; the wrapper throws an error if you try to call it and it didn't load.

File: source/codegen/templates/library.cpp.mako (hunk 1 above)
Comment (astarche) [reply source/codegen/templates/library.cpp.mako]: Like Marcos, I was expecting something like:

```c++
  if (function_pointers_.SetRuntimeEnvironment) {
    function_pointers_.SetRuntimeEnvironment(environment, environmentVersion, reserved1, reserved2);
  }
```

But I agree that try/catch is not the end of the world.

I'm always a fan of adding metadata conventions. But I think there's already a strong convention in the drivers to implement/use `SetRuntimeEnvironment` in this way and that it's implemented similarly in `nimi-python` so I'm happy for the team to use their judgement on it.

Is it possible that we'd want to automatically add `SetRuntimeEnvironment` even if drivers don't declare it in their metadata? If the convention is strong enough, it can be better to have it be global.

Comment (ni-jfitzger) [reply source/codegen/templates/library.cpp.mako]: I actually just added that check.
Let me know if you think I need to document the reason for the call.

File: generated/version.h (hunk 4 above)
Comment (ni-jfitzger) [reply generated/version.h]: Why the `static`? Since it's a constexpr, won't we just evaluate it at compile time and then not need it anymore?

File: source/server/version.h.in [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace nidevice_grpc {
+
+std::string const NIDEVICE_GRPC_MAJORNUMBER = "@PROJECT_VERSION_MAJOR@";
+std::string const NIDEVICE_GRPC_MINORNUMBER = "@PROJECT_VERSION_MINOR@";
+std::string const NIDEVICE_GRPC_BUILDNUMBER = "@PROJECT_VERSION_PATCH@";
+std::string const NIDEVICE_GRPC_MODIFICATIONNUMBER = "0";
+std::string const NIDEVICE_GRPC_FILEVERSION = "@PROJECT_VERSION_MAJOR@.@PROJECT_VERSION_MINOR@.@PROJECT_VERSION_PATCH@.0";
+std::string const NIDEVICE_GRPC_PRODUCTVERSION = "@PROJECT_VERSION_MAJOR@.@PROJECT_VERSION_MINOR@";
+std::string const NIDEVICE_GRPC_ORIGINALFILENAME = "ni_grpc_device_server.exe";
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/server/version.h.in]: Done.

File: source/server/version.h.in (hunk 7 above)
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/server/version.h.in]: Updated to use Google style convention.

File: generated/version.h (hunk 4 above)
Comment (bkeryan) [reply generated/version.h]: For a variable at file/namespace scope, [`static` means "internal linkage"](https://en.cppreference.com/w/cpp/language/storage_duration), so each translation unit (.cpp file) basically gets a copy, preventing name collisions between translation units. I don't know if it's necessary for `constexpr`. In C++, `const` variables are static by default and non-`const` variables are global by default. (AFAIK C has different rules.)

`constexpr` means the variable can be used at compile time, not that the variable cannot exist at run time.

**Review by bkeryan — APPROVED** [review]

File: source/codegen/metadata/nidcpower/__init__.py (hunk 2 above)
Comment (reckenro) [reply source/codegen/metadata/nidcpower/__init__.py:16]: Yeah, I think this should be fine. May need to come back to it if we start using override_metadata for more things and it gets more complex, but not sure on that and don't think it's worth pre-emptively addressing.

Historically, we've only used our _addon.py files for metadata_validation suppressions for certain things and not for overriding any of the main metadata: https://github.com/ni/grpc-device/blob/main/source/codegen/metadata/nifgen/enums_addon.py#L7

And going forward, in general I don't think we'd want to modify existing metadata through this mechanism but would rather just fix the source of it (fix generation of it elsewhere or hand-edit it in the regular `functions.py`).

**Review by astarche — APPROVED** [review]

File: source/codegen/templates/library.cpp.mako [hunk 11]
```diff
@@ -21,6 +21,7 @@ windows_library_name = windows_library_info['name']
 // Service implementation for the ${config["driver_name"]} Metadata
 //---------------------------------------------------------------------
 #include "${module_name}_library.h"
+#include "version.h"
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/codegen/templates/library.cpp.mako:24]: Yeah, I can do that. I only didn't to start with because I wasn't sure you would want the extra template complexity.

**Review by reckenro — APPROVED** [review]

File: source/codegen/templates/library.cpp.mako (hunk 11 above)
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/codegen/templates/library.cpp.mako:25]: Done.

File: source/server/version.h.in (hunk 8 above)
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/server/version.h.in]: Used `static constexpr char*`, based on this thread and another.

File: generated/nidcpower/nidcpower_library.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply generated/nidcpower/nidcpower_library.cpp:1378]: Made the call conditional.

File: source/codegen/templates/library.cpp.mako (hunk 1 above)
Comment (ni-jfitzger) [reply source/codegen/templates/library.cpp.mako]: >Rather than adding specific logic for a specific function, it might be better to define this requirement in the metadata and use that to add any extra "initialization functions" so that this is extensible to other functions in the future. I'll defer to @reckenro or @astarche on that since they know more about the code generation, though.

I'll stick with the YAGNI (You Ain't Gonna Need It) principle and avoid overcomplicating it, since I haven't been told other wise.

File: source/codegen/templates/library.cpp.mako [hunk 12]
```diff
@@ -44,6 +47,12 @@ ${service_class_prefix}Library::${service_class_prefix}Library() : shared_librar
 %>\
   function_pointers_.${method_name} = reinterpret_cast<${method_name}Ptr>(shared_library_.get_function_pointer("${c_name}"));
 % endfor
+% if 'SetRuntimeEnvironment' in service_helpers.filter_api_functions(functions, only_mockable_functions=False):
+
+  if (function_pointers_.SetRuntimeEnvironment) {
+    this->SetRuntimeEnvironment(nidevice_grpc::kNiDeviceGrpcFileVersion, nidevice_grpc::kNiDeviceGrpcOriginalFileName, "", "");
```
[the remark this answers is not in the document]
Comment (ni-jfitzger) [reply source/codegen/templates/library.cpp.mako]: Fixed.

File: generated/nidcpower/nidcpower_library.h (hunk 5 above)
Comment (ni-jfitzger) [reply generated/nidcpower/nidcpower_library.h:169]: Created https://github.com/ni/grpc-device/pull/965

**Comment by ni-jfitzger** [comment]
I've been thinking about how to test that SetRuntimeEnvironment is called when it should be and not called when it shouldn't be.

* A mock of the NiFakeLibrary constructor (where the SetRuntimeEnvironment call happens) won't do anything except what the caller tells it to do.
* If I move the SetRuntimeEnvironment call up a level into the function that calls the constructor, well that function creates the library object, itself, so I can't pass in a mock of the Library class.
* The SharedLibrary object that loads the dll is constructed by the NiFakeLibrary constructor, so I can't pass in a fake or mock of SharedLibrary to noop the library loading.

@reckenro @astarche How would you feel about me tweaking the Library classes to have the SharedLibrary type passed to the constructor (so that a test could pass a mock)?

**Comment by astarche** [comment]
@ni-jfitzger @reckenro Passing in the shared library sounds alright to me. I believe you could make it a `unique_ptr` and pass it in via `make_unique`. A shared_library abstraction is a reasonable place to introduce a pointer for testability.

Seems more likely that this code would fail because `shared_library` doesn't do what you expect rather than because your `if` statement doesn't do what you expect. But we haven't invested in building some dummy dll for testing `shared_library` so now is not the time to do it. We typically fall back to relying on driver-level acceptance tests, but in this case I guess we don't have a good way to test that either (both hard to verify the call and hard to find a driver that doesn't support it).

So overall: sounds good.

**Comment by ni-jfitzger** [comment]
This PR is being abandonded in favor of #973

**Comment by bkeryan** [comment]
> Passing in the shared library sounds alright to me. I believe you could make it a unique_ptr and pass it in via make_unique.

Note that the `NifooLibrary` constructor will move the test's `MockSharedLibrary`, invalidating it. shared_ptr may be simpler, especially if you need to access the mock after the `NifooLibrary` takes ownership.

> But we haven't invested in building some dummy dll for testing shared_library so now is not the time to do it.

With mocking, you don't need a dummy DLL. You can set up the `MockSharedLibrary` to return pointers to functions implemented in the test EXE. However, it still seems like a relatively low value test.

> We typically fall back to relying on driver-level acceptance tests, but in this case I guess we don't have a good way to test that either (both hard to verify the call and hard to find a driver that doesn't support it).

Is it hard to write a C++ test that logs ETW events? In C#, we have test utilities for writing tests that capture ETW events for CEIP or whatever.
